### PR TITLE
Mail transport abstraction

### PR DESCRIPTION
**Pull request by gonzaloebiz**
We changed the constructor on lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php so its possible to have $message as an interface  \Magento\Framework\Mail\MessageInterface, before it was an instance of \Magento\Framework\Mail\Message which we consider wrong.

Also we removed all methods that dont belong to the transport: addCc, addTo, addBcc, setReplyTo and setFrom, we consider these message operations.

We added a getter and one setter so its possible to change the message via an interceptor on the fly.

On lib/internal/Magento/Framework/Mail/TransportInterfaceFactory.php we added a getter for $_instanceName, this allows an interceptor to change the transport instead of needing a rewrite for this.

### CODE REVIEW

**Comment by centerax** [comment]
We KNOW that the tests will fail at this point, just need to validate with you (Magento) and the Community that this solution is the way to go.

**Comment by fooman** [comment]
I am liking the direction of this. 

The duplication of addCc et al didn't make sense to me either.

It appears there is some unfinished work in the existing code around setFrom - the Interface uses `setFrom($fromAddress)` vs `setFrom($email, $name = null). We also have a SenderResolver but we don't do the same on other email addresses.

I totally agree that a getMessage() on the transport is a very good idea.

Hope to see improvements going into the mail sending code. It certainly is an area where Magento1 struggles with multiple extension required to rewrite the same code to achieve additional functionality.

**Comment by centerax** [comment]
Thanks for your feedback Kristof (@fooman), we are really looking forward to Magento's feedback on this since as you say Magento1 fails short on this subject.

I think that is the spirit of all developers contributing to Magento2 to get things right from scratch.

**Comment by choukalos** [comment]
Talked to Rick.  He's going to assign an architect to look into the PR and provide feedback.  Thanks a ton for doing this.  Have solid interfaces in the system (and the system refactored to use those interfaces) is the right direction from the product perspective.  In this case it looks like you're effectively creating an API that email providers would implement.  Assuming architecture believes this is stable go forward that'll make it easy to have a set of extensions for various email providers.  That helps check off one of the items on my backlog!

**Comment by centerax** [comment]
Thanks for the update Chuck!

File: lib/internal/Magento/Framework/Mail/TransportInterfaceFactory.php [hunk 1]
```diff
@@ -49,4 +49,14 @@ public function create(array $data = [])
     {
         return $this->_objectManager->create($this->_instanceName, $data);
     }
+
+    /**
+     * @param $name
+     * @return $this
+     */
+    public function setInstanceName($name)
```
Comment (TexanHogman) [inline lib/internal/Magento/Framework/Mail/TransportInterfaceFactory.php]: I am surprised this factory exists in the codebase as I expect all these to be code generated; I suspect this is a bug will get confirmation from @antonkril -- regardless you should specify a di preference to your own factory in TransportBuilder and not requiring this to be done via a plugin.

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php [hunk 2]
```diff
@@ -75,14 +75,14 @@ class TransportBuilder
...
-     * @param \Magento\Framework\Mail\Message $message
+     * @param \Magento\Framework\Mail\MessageInterface $message
      * @param \Magento\Framework\Mail\Template\SenderResolverInterface $senderResolver
      * @param \Magento\Framework\ObjectManagerInterface $objectManager
      * @param \Magento\Framework\Mail\TransportInterfaceFactory $mailTransportFactory
      */
     public function __construct(
         \Magento\Framework\Mail\Template\FactoryInterface $templateFactory,
-        \Magento\Framework\Mail\Message $message,
+        \Magento\Framework\Mail\MessageInterface $message,
```
Comment (TexanHogman) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: with the removal of all the delivery setters for shouldn't be needed as construct arg.

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php [hunk 3]
```diff
@@ -95,69 +95,22 @@ public function __construct(
...
-
-    /**
-     * Add bcc address
-     *
-     * @param array|string $address
-     * @return $this
+     * @return \Magento\Framework\Mail\MessageInterface
      */
-    public function addBcc($address)
+    public function getMessage()
```
Comment (TexanHogman) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: don't see a reason to provide a getter on the message

Comment (fooman) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: I am currently following these discussions with interest from the sideline. I am looking at ways of how to best port my EmailAttachments extension to Magento 2. In Magento 1 I used a rewrite on the Template Mailer https://github.com/fooman/emailattachments/blob/master/app/code/community/Fooman/EmailAttachments/Model/Core/Email/Template/Mailer.php#L54 to in the end allow me to add an attachment to the message. Obviously being able to do the same without a rewrite (di preference) would be great.

My current reading of the Magento2 code keeps the message object very private. So far I haven't seen a good entry point to access the message to be able to manipulate more than the template content. A getMessage() to access the message object for further manipulation seems a good starting point but maybe there is a better way which I haven't come across yet.

File: lib/internal/Magento/Framework/Mail/TransportInterfaceFactory.php (hunk 1 above)
Comment (antonkril) [inline lib/internal/Magento/Framework/Mail/TransportInterfaceFactory.php]: Agree with Rick. Making such call from plugins doesn't make it non-conflicting. Preferences are a standard declarative way of doing conflicting modifications. It's easier to track them to find conflicts between modules.

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php (hunk 2 above)
Comment (antonkril) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: Message is non-injectable. Shouldn't be constructor argument. But it wasn't added in this PR

**Comment by antonkril** [comment]
Are you planning to you address the comments in review?

**Comment by centerax** [comment]
Sorry for the delay @antonkril , we will address this this week.

**Comment by vpelipenko** [comment]
We are waiting for updates from you.

**Comment by gonzaloebiz** [comment]
I know the DI is the better method to change the preferred implementation of an interface. But this method i think is a quite dangerous for core functions. Suppose i change the preferred implementation with a DI, then the user not activate the module in the admin, i can't control this situation in my implementation i think this is not correct.
Suppose other provider install your own preferred implementation via DI, then our extension will stop working
The point is, how i can add a new transport without interfere with the whole email implementation and affect future extensions?

**Comment by antonkril** [comment]
You can create proxy implementation of transport interface that will check in config if your extension is active and call to your implementation if true and to default otherwise.

**Comment by centerax** [comment]
@antonkril Do you think that is the optimal solution?
What happens if our extension and for example @fooman 's do the same (using DI)?
What happens then?

Our goal with this pull request is to start the discussion and make things "right" from scratch since "we" are building the commerce platform for the future. I don't think its a good thing to release Magento2 to merchants with this still not properly addressed.

Now, if someone can explain with a solid foundation why this change is not a good thing we will let it be and do things as with M1.

Any further feedback would be appreciated.

Thanks.

**Comment by antonkril** [comment]
There are 2 types of changes you can do: extension and modification.
Extension is a non-conflicting change. You can do it with events or plugins.

But modification is inherently conflicting change. When you want to introduce new DB adapter, this is a conflicting change because you don't want to write to two adapters, you need only one. So you have to substitute core DB adapter with yours. And if some other dev creates his own Adapter then he's in conflict with you. We can not do anything about this. He will have to substitute your adapter.

**Comment by gonzaloebiz** [comment]
@antonkril
as you can see at https://github.com/magento/magento2/blob/develop/lib/internal/Magento/Framework/Mail/Transport.php#L9 the transport is strongly linked to Zend_Mail_Transport_Sendmail
The only way i see to make the change via plugin, is to have the posibbility to change the instanceName (https://github.com/ebizmarts/magento2/blob/develop/lib/internal/Magento/Framework/Mail/TransportInterfaceFactory.php#L57) This way we (and any other provider) can add transporters

**Comment by antonkril** [comment]
You don't need a plugin or a setter here. You can use interface preference for TransportInterface, since this is a conflicting modification anyway.

**Comment by snohe** [comment]
@gonzaloebiz, we haven't heard back from you for quite some time.  Can you let us know if you are still working on this?

**Comment by vpelipenko** [comment]
Closed. There is no response from contributor. @gonzaloebiz, if you have any updates and still work on this pull request, please, feel free to reopen it.

**Pull request closed without merging**